Add announcement document (#67)
larry-robotics/elkodon: ANNOUNCEMENT.md

to evaluate whether Rust truly lived up to its promises. The resounding answer is a
definitive "yes," and more.

My day job involves certifying iceoryx written in C++ for use in safety-critical
ASIL-D environments. While grappling with algorithmic challenges, the burden of C++
introduces a myriad of additional complications, consuming considerable time and
causing frustration.

* **Documentation:** Managing doxygen code examples in harmony with the codebase.
* **Lifetimes:** Navigating lifetime dependencies when working with RAII and factories.
* **Concurrency:** Preventing inadvertent multi-threaded usage and certifying
    concurrent and lock-free code.
* **Build-System:** Handling dependencies and third-party packages seamlessly.
* **Templates:** Certifying generic C++ code for various but valid types.
* **Static Code Analysis:** Choosing the right tools and adhering to Misra,
    Autosar, C++ Core Guidelines, or a combination.

Why dwell in the abstract when we can scrutinize a real-world example, comparing
iceoryx (C++) to iceoryx2 (Rust)? Brace yourself for a journey into a future where
Rust takes the lead in inter-process zero-copy communication within our specialized

> domain. Iceoryx2 is not just an upgrade—it's a leap forward into a safer, more

Reviewer: ```suggestion
domain. Iceoryx2 is not just an upgrade — it's a leap forward into a safer, more
```

In this context, the `Publisher` assumes the role of the sender in a publish-subscribe 
messaging pattern. When users intend to transmit data, they typically follow a set
of common steps across both versions.

1. Invoke `Publisher::loan()` to obtain a `Sample` for storing the data to be sent.
2. Utilize `Publisher::send(sample)` to dispatch the data to all receiving endpoints (`Subscriber`).

Represented here is a simplified interface for this `Publisher`:

```cpp
// C++
class Publisher {
  public:
    Sample loan();
    void send(Sample &&sample);
};
```

```rust
// Rust
struct Publisher {}

impl Publisher {
    fn loan(&'publisher self) -> Sample<'publisher> {}
    fn send(&self, sample: Sample) {}
}
```

## Documentation Best Practices: C++ vs. Rust

When it comes to documenting code and ensuring the accuracy of code examples, both C++
and Rust offer distinct approaches. Let's delve into the practices and tools used in
each language:

### C++

In the C++ realm, internal documentation is commonly managed using Doxygen, a tool
providing `@code`/`@endcode` tags for embedding code examples. However, there's a

> crucial limitation—Doxygen does not verify whether the code is running or compiling.

Reviewer: ```suggestion
crucial limitation — Doxygen does not verify whether the code is running or compiling.
```

This introduces a potential risk, as refactoring in one part of the codebase may
inadvertently break examples located elsewhere.

```cpp
/// @code
///   std::cout << "hello world, without include iostream";

> /// @endcode

Reviewer: We don't make use of it in iceoryx but there is an option to embed code snippets from a C++ file. This prevents the issue with non-compiling examples. But one needs to look at the doxygen output to check whether the examples is still semantically correct. With Rust it's just at the start of the method and easier to check in PRs

the `Send` trait, which is not implemented by default. Any attempt to violate 
this contract is again caught by the compiler during compilation, demonstrating 
Rust's ability to handle such problems at compile time without introducing 
performance or implementation overhead.

## Future

This leads me back to the initial question: Why does the automotive domain lag 
behind?

A mere year ago, when I embarked on the development of iceoryx2, the absence of 
a certified Rust compiler posed a critical barrier. Without this fundamental 
component, certifying code in the automotive domain seemed nearly insurmountable. 
Additionally, projects within the automotive sector often span years and 
frequently involve interactions with legacy C and C++ code bases—essential 
factors that cannot be overlooked.

Today, however, the landscape has evolved significantly. We now have access to 
a [certified Rust compiler](https://ferrous-systems.com/ferrocene/). Rust 
facilitates seamless invocation of functions from C code bases, thanks to 
[bindgen](https://github.com/rust-lang/rust-bindgen), and although incorporating 
C++ code poses some challenges, it is achievable. Moreover, the reverse is also 

> true—integrating a Rust library into an existing C/C++ code base and invoking 

Reviewer: ```suggestion
true — integrating a Rust library into an existing C/C++ code base and invoking 
```

[Rust functions within C](https://doc.rust-lang.org/nomicon/ffi.html) is feasible.

> In my estimation, the opportune moment has arrived to take the next stride and 

Reviewer: ```suggestion
In my estimation, the window of opportunity has arrived to take the next stride and 
```

emerge as pioneers in successfully certifying code for ISO 26262 (ASIL D), with 
iceoryx2 as the ideal project for this endeavor. While this undertaking will 
undoubtedly require time, particularly given the intricacies of inter-process 
zero-copy communication, I approach it with confidence. Drawing on my experience 
certifying C++ code and the insights gained from rewriting iceoryx in Rust, I 
believe we can navigate these challenges more efficiently.

Embracing Rust not only accelerates the development of new software components 
but also reduces costs, courtesy of the modern safety and developer features it 

> offers. This, in turn, empowers our customers to write superior code swiftly, 

Reviewer: ```suggestion
offers. This, in turn, empowers our users to write superior code swiftly, 
```
Since this is an open source project `Users` might suite better
